resources/channel.sub.js

```diff
@@ -0,0 +1,1003 @@
+                        flags: item.flags
+                    };
+                    break;
+                case "date":
+                    value = Date.prototype.toJSON.call(item);
+                    break;
+                case "error":
+                    value = {
+                        type: item.constructor.name,
+                        name: item.name,
+                        message: item.message,
+                        lineNumber: item.lineNumber,
+                        columnNumber: item.columnNumber,
+                        fileName: item.fileName,
+                        stack: item.stack,
+                    };
+                    break;
+                case "array":
+                case "set":
+                    value = [];
+                    newTarget = {type: "array", value};
+                    for (let child of item) {
+                        stack.push({item: child, target: newTarget});
+                    }
+                    break;
+                case "object":
+                    value = {};
+                    newTarget = {type: "object", value};
+                    for (let [targetKey, child] of Object.entries(item)) {
+                        stack.push({item: child, target: newTarget, targetKey});
+                    }
+                    break;
+                case "map":
+                    value = [];
+                    newTarget = {type: "map", value};
+                    for (let [targetKey, child] of item.entries()) {
+                        stack.push({item: targetKey, target: newTarget, targetKey: "key"});
+                        stack.push({item: child, target: newTarget, targetKey: "value"});
+                    }
+                    break;
+                default:
+                    throw new TypeError(`Can't serialize value of type ${type}; consider using RemoteObject.from() to wrap the object`);
+                };
+            }
+            let result = {type};
+            if (value !== undefined) {
+                result.value = value;
+            }
+            if (objectId !== undefined) {
+                result.objectId = objectId;
+            } else {
+                objectsSeen.set(item, result);
+            }
+            if (!target) {
+                if (serialized !== null) {
+                    throw new Error("Tried to create multiple output values");
+                }
+                serialized = result;
+            } else {
+                switch (target.type) {
+                case "array":
+                    target.value.push(result);
+                    break;
+                case "object":
+                    target.value[targetKey] = result;
+                    break;
+                case "map":
+                    if (targetKey === "key") {
+                        target.value.push([]);
+                    }
+                    target.value[target.value.length - 1].push(result);
```

Comment:
This needs a comment, it took me a lot of time to unpack. It depend on the traversal order given by the queue, where we have to get "key" first, and the last line here is used for both the "key" and "value" case. So we do end up pushing to the new `[]` twice, in the right order.